Add support for fan preset modes

mill1000 opened the pull request:
# What does this implement/fix?

Adds preset mode support to fan components.  Requires  https://github.com/esphome/aioesphomeapi/pull/616 and for Home Assistant users: https://github.com/home-assistant/core/pull/103781

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes https://github.com/esphome/feature-requests/issues/1661

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3362

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# For testing speed fan presets
fan:
  - platform: speed
    name: "Test Fan"
    id: test_fan
    output: gpio_19
    oscillation_output: gpio_18
    direction_output: gpio_17
    preset_modes:
      - Auto
      - Sleep
      - Some Other Preset
    on_preset_set:
      then:
        lambda: |-
          if (id(test_fan).preset_mode == "Auto") {
            // Auto enabled, do automatic things
          }
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

mill1000 committed:
Update protobufs for fan preset modes

mill1000 committed:
Update FanTrains for preset modes

mill1000 committed:
Add preset_mode member and accessors to base fan class

mill1000 committed:
Update API with fan preset modes

mill1000 committed:
Update copy_fan for preset modes

mill1000 committed:
Fix clang-tidy issues

mill1000 committed:
Fix issues with FanRestoreState due to std::string being non-trivial.

Instead store an index into the preset array

mill1000 committed:
Add a helper method `supports_preset_modes` to reduce some verbosity

mill1000 committed:
Add preset support to Speed Fan component.

Each preset mode can specify a speed, direction and oscilatting setting to apply.

Users can create as many presets as they like but the names must be unique.

mill1000 committed:
Fix bad logic on control call that caused states to not be published

mill1000 committed:
Fix lint issues and use the correct formatter

mill1000 commented:
Embarrassed that I keep jumping the gun on my pushes and not passing CI...

mill1000 committed:
Check for preset_modes before trying to interate over them..

bdraco commented:
I've only had time to do a cursory look through but I wanted to make sure to point out the below:

https://developers.home-assistant.io/docs/core/entity/fan?_highlight=fan#preset-modes

> Preset modes should not include named (manual) speed settings as these should be represented as percentages.

> Manually setting a speed must disable any set preset mode. If it is possible to set a percentage speed manually without disabling the preset mode, create a switch or service to represent the mode.


You may have already implemented it this way, but I wanted to make sure I mentioned it now since it may save some time later

mill1000 commented:
> https://developers.home-assistant.io/docs/core/entity/fan?_highlight=fan#preset-modes
> 
> > Preset modes should not include named (manual) speed settings as these should be represented as percentages.

Hm...not sure how to feel about this one. Was the intent to prohibit presets that affect speed entirely, or to prevent an integration from using presets as the sole means to control speed. e.g. Is it unreasonable to want 1 % granularity but have presets for specific speeds (i.e. Low, Med, High)

> > Manually setting a speed must disable any set preset mode. If it is possible to set a percentage speed manually without disabling the preset mode, create a switch or service to represent the mode.
> 
> You may have already implemented it this way, but I wanted to make sure I mentioned it now since it may save some time later

As currently implemented (in speed fan) this is true. Any command to change the speed, direction or oscillation will disable the preset.

mill1000 commented:
Poked around with adding preset support to H-Bridge fan as well, but holding off on any continued work until it's decided whether preset modes with speeds are not allowed.

mill1000 commented on esphome/components/copy/fan/copy_fan.cpp:
```diff
@@ -46,6 +49,8 @@ void CopyFan::control(const fan::FanCall &call) {
   if (call.get_direction().has_value())
     call2.set_direction(*call.get_direction());
+  if (!call.get_preset_mode().empty())
+    call2.set_preset_mode(call.get_preset_mode());
```
I think this logic is incorrect. Copy fan should probably always pass preset mode, even if empty.

mill1000 replied:
On second thought this is OK as in. `call2` will default to an empty `preset_mode` so there's no need to copy an empty `preset_mode` from `call` to `call2`

bdraco commented:
> Hm...not sure how to feel about this one. Was the intent to prohibit presets that affect speed entirely, or to prevent an integration from using presets as the sole means to control speed. e.g. Is it unreasonable to want 1 % granularity but have presets for specific speeds (i.e. Low, Med, High)

The intent was to make sure that presets are not simple speeds like `Low` `Low Medium` `Med` `Medium High` `High` etc as these should be handled by percentage.  The goal is to model a fan so that when you ask for 20% between different types of fans you get the nearest to 20% as possible.

Preset modes should be something like `Auto`, `Breeze`, `Cycle`, these may control the speed but it should not be a 1:1 mapping to a specific speed

mill1000 commented:
So what do you feel is the path forward here? Does allowing control of direction and oscillation ensure that "presets are not simple speeds", or should I rip out the speed fan implementation?

bdraco commented:
> So what do you feel is the path forward here? Does allowing control of direction and oscillation ensure that "presets are not simple speeds", or should I rip out the speed fan implementation?

I think its ok. ESPHome fans are going to be mostly DIY, and I don't think it makes it any easier for some vendor to build an ESPHome fan and only `low`,`medium`,`high` presets modes instead of speeds.

mill1000 commented:
Cool. I'll continue with the h-bridge work and additional test cases.

mill1000 committed:
Add preset modes to H-bridge fan

mill1000 committed:
Add some compile tests for Speed and Hbridge fans with presets

nagyrobi commented:
Are the values templatable?

My use case: my fans can be controlled 0-100% to any value. However, having to drag a small slider is unconfortable so simply having additional presets with say 33%, 66%, and 100% as shortcuts (names could vary) would improve the user experience of the family nembers. 

Templating could be handy to have the ability to change the values at runtime. Eg have a number component to specify the actual speed of `Breeze` from the ui, rather than having to reflash firmware.

bdraco commented:
> @bdraco my use case: my fans can be controlled 0-100% to any value. However, having to drag a small slider is unconfortable so simply having additional presets with say 33%, 66%, and 100% as shortcuts (names could vary) would improve the user experience of the family nembers.

I get what you are saying, but that is a frontend problem. We shouldn't let frontend design dictate how the backend works, we should improve the frontend instead.

nagyrobi commented:
Why a frontend problem? We would want to use both: presets for quick speed access, and the slider too, for special occasions.

bdraco commented:
> Why a frontend problem? We would want to use both: presets for quick speed access, and the slider too, for special occasions which need intermediate levels.

We shouldn't add multiple ways to do the same thing on the backend as we don't want to have to maintain multiple systems to do the same thing.

There are already frontend options to set a preset speed, one of them is a button card:

```yaml
show_name: true
show_icon: true
type: button
entity: fan.alexander_master_bedroom_2
tap_action:
  action: call-service
  service: fan.set_percentage
  target:
    entity_id: fan.alexander_master_bedroom_2
  data:
    percentage: 18
icon: mdi:fan-speed-2
```

nagyrobi commented:
@bdraco the only issue with this approach is that it only considers Home Assistant's UI as the single usage possibility with ESPHome. There are also other open source user interfaces which rely on HA core, but don't need a PC/browser to control it (see https://www.openhasp.com for example). Also ESPHome is supported (even at api level) by other open source systems, like iobroker for example, which is a good thing.

bdraco commented:
> @bdraco the only issue with this approach is that it only considers Home Assistant's UI as the single usage possibility with ESPHome. There are also other open source user interfaces which rely on HA core, but don't need a PC/browser to control it (see https://www.openhasp.com for example). Also ESPHome is supported (even at api level) by other open source systems, like iobroker for example, which is a good thing.

This PR is proposing to add support to Home Assistant so I'm reviewing it in that context.

mill1000 commented:
Ok. I've tested the preset mode on hbridge, speed and copy fan. At this time I'm satisfied with the function and have marked this as ready for review

bdraco commented on esphome/components/fan/fan.cpp:
```diff
@@ -143,20 +173,36 @@ void Fan::save_state_() {
+  if (this->get_traits().supports_preset_modes() && !this->preset_mode.empty()) {
+    const auto &preset_modes = this->get_traits().supported_preset_modes();
+    // Store index of current preset mode
+    auto it = std::find(preset_modes.begin(), preset_modes.end(), this->preset_mode);
+    if (it != preset_modes.end())
```
Please name variables to show intent

bdraco commented:
The one thing that looks missing is that when the user manually sets a speed, it should override and clear the preset mode

mill1000 committed:
Address review comment about iterator name

mill1000 committed:
Change logic around applying presets to avoid need for ignore_preset arg

mill1000 commented on esphome/components/speed/fan/speed_fan.cpp:
```diff
@@ -27,9 +36,18 @@ void SpeedFan::control(const fan::FanCall &call) {
   if (call.get_direction().has_value())
     this->direction = *call.get_direction();
 
+  if (!ignore_preset) {
```
I don't love this logic and I've worked up a different implementation, but I still need to test it

mill1000 replied:
How's the name `preset_iterator`. Wasn't sure what else to call it.

mill1000 commented:
> The one big thing that looks missing is that when the user manually sets a speed, it should override and clear the preset mode

In my testing this has worked as you've described. From HA, any manual change to speed, direction or oscillation causes a call to be made with an empty `preset_mode` which clears the preset.

bdraco commented:
> > The one big thing that looks missing is that when the user manually sets a speed, it should override and clear the preset mode
> 
> In my testing this has worked as you've described. From HA, any manual change to speed, direction or oscillation causes a call to be made with an empty `preset_mode` which clears the preset.

👍  I was looking for the code here, and didn't think about HA will send it empty.

I don't have a fan running ESPHome unfortunately

mill1000 commented:
@nagyrobi 

> Are the values templatable?
> 
> Templating could be handy to have the ability to change the values at runtime. Eg have a template number component to specify the actual speed of Breeze with a lambda (thus from the ui), rather than having to reflash firmware.

No they aren't templatable. 

I looked into this briefly, and I don't _think_ there's a technical limitation to templating them but I wasn't able to find another component that used templates outside of actions (except of course template components themselves).

bdraco commented:
I didn't get a chance to test this out yet and I'm traveling again tomorrow but I'm going to throw the test hardware in my travel bag so I'll be able to test it on the road

bdraco committed:
Merge branch 'dev' into feature/fan_presets

bdraco commented:
Tested speed fan and everything is working as expected AFAICT

mill1000 commented:
Sweet!

filmkorn commented on esphome/components/hbridge/fan/__init__.py:
```diff
@@ -28,6 +37,19 @@
+                cv.Optional(CONF_SPEED): cv.int_range(1),
+                cv.Optional(CONF_DIRECTION): cv.enum(FAN_DIRECTION_ENUM, upper=True),
+            },
+            cv.has_at_least_one_key(CONF_SPEED, CONF_DIRECTION),
```
To implement an `auto` mode that controls the fan speed (and perhaps direction) based on other factors (ie air pollution or temperature), it seems obsolete to set at least speed and direction. In this case I'd want the preset name to be nothing but an enum that I can use as input and trigger for an automation.

bdraco replied:
It seems like driving the preset mode via an automation might be a better design where the preset mode effectively turns on an automation and setting manual speed or changing presets will turn it off

mill1000 replied:
Either way for that to be possible we'd need a new automation trigger, something like `on_preset_set`

bdraco replied:
That makes sense.  I really like the idea that the preset mode drives an automation as it's much more flexible and inline with what we think of as being a preset mode

mill1000 replied:
Alright I'll take a swing at it

mill1000 replied:
Added a new `on_preset_set` trigger

```yaml
    on_preset_set:
      then:
        - logger.log: "A preset was set"
````

mill1000 committed:
Add on_preset_set trigger to allow automation based on preset mode.

Remove requirement for preset modes to specify at least a speed, direction or oscillation.

nagyrobi commented:
Within the automation, you detect which preset was set by name? Name can be any string (spaces, diacritics, etc)?

bdraco commented on esphome/components/speed/fan/__init__.py:
```diff
@@ -1,18 +1,40 @@
 
 SpeedFan = speed_ns.class_("SpeedFan", cg.Component, fan.Fan)
 
+_PRESET_MODES_SCHEMA = cv.All(
+    cv.ensure_list(
+        {
+            cv.Required(CONF_NAME): cv.string_strict,
+            cv.Optional(CONF_SPEED): cv.int_range(1),
+            cv.Optional(CONF_OSCILLATING): cv.boolean,
+            cv.Optional(CONF_DIRECTION): cv.enum(FAN_DIRECTION_ENUM, upper=True),
+        }
+    ),
+    cv.Length(min=1),
+)
```
Since preset modes are an advanced feature, I think it would be better to not have to support both manually configured and the automation.  They can do all of this in the automation anyways.

mill1000 replied:
Gotcha. Just pushed an automation-only approach.

Presets are now defined as a list of names
```yaml
preset_modes:
  - Auto
  - Sleep
```

mill1000 committed:
Rework fan preset modes to support an automation-only approach.

- Preset modes are now defined as a list of unique names without any speed, or other values
- Users are expected to define an on_preset_set trigger in order to implement actualy preset function.

mill1000 committed:
Fix tests for new preset mode schema

mill1000 committed:
Fix clang-tidy

bdraco committed:
Merge branch 'dev' into feature/fan_presets

mill1000 commented:
@bdraco Anything else you need from me to move this along?

bdraco commented:
Nope, at this point I'm waiting for another dev to take a look

jesserockz reviewed (approved):
Code looks fine, but what do presets actually do on a fan? Doesn't seem like there is any functional change?

mill1000 commented:
> but what do presets actually do on a fan? Doesn't seem like there is any functional change?

For speed fan and h-bridge they do nothing without configuring a matching `on_preset_set` automation.

Other platforms might have predefined presets to enable device specific modes. (e.g. an automatic or sleep mode).
